metagraph/src/graph/representation/hash/dbg_sshash.cpp

```diff
@@ -0,0 +1,356 @@
+#include "kmer/kmer_extractor.hpp"
+
+
+namespace mtg {
+namespace graph {
+
+#if _PROTEIN_GRAPH
+static constexpr uint16_t bits_per_char = sshash::aa_uint_kmer_t<uint64_t>::bits_per_char;
+#else
+static constexpr uint16_t bits_per_char = sshash::dna_uint_kmer_t<uint64_t>::bits_per_char;
+#endif
+
+template <typename T>
+struct template_parameter;
+
+template <template <typename ...> class C, typename T>
+struct template_parameter<C<T>> { using type = T; };
+
+template <typename T>
+using get_kmer_t = typename template_parameter<typename std::remove_const<typename std::remove_reference<T>::type>::type>::type;
+
+const std::string DBGSSHash::alphabet_ = kmer::KmerExtractor2Bit().alphabet;
+
+constexpr DeBruijnGraph::node_index sshash_to_graph_index(uint64_t idx) {
+    return idx + 1;
+}
+constexpr uint64_t graph_index_to_sshash(DeBruijnGraph::node_index idx) {
+    return idx - 1;
+}
+
+size_t DBGSSHash::dict_size() const {
+    return std::visit([](const auto &d) { return d.size(); }, dict_);
+}
+
+DBGSSHash::DBGSSHash(size_t k, Mode mode) : k_(k), num_nodes_(0), mode_(mode) {
+    size_t odd_k = (k_ | 1);
+
+    if (odd_k * bits_per_char <= 64) {
+        dict_ = sshash::dictionary<kmer_t<KmerInt64>>();
+    } else if (odd_k * bits_per_char <= 128) {
+        dict_ = sshash::dictionary<kmer_t<KmerInt128>>();
+    } else if (odd_k * bits_per_char <= 256) {
+        dict_ = sshash::dictionary<kmer_t<KmerInt256>>();
+    } else {
+        common::logger->error("k too big: {} > {}", odd_k, 256 / bits_per_char);
+        throw std::length_error("k fail");
+    }
+}
+
+DBGSSHash::DBGSSHash(std::string const& input_filename, size_t k, Mode mode)
+      : DBGSSHash(k, mode) {
+    sshash::build_configuration build_config;
+    build_config.k = k;
+    // quick fix for value of m... k/2 but odd
+    build_config.m = std::min(uint64_t(k / 2) | 1, sshash::constants::max_m);
+    build_config.verbose = common::get_verbose();
+    build_config.num_threads = get_num_threads();
+
+    // silence sshash construction messages when not verbose
+    if (!common::get_verbose())
+        std::cout.setstate(std::ios_base::failbit);
+
+    std::visit([&](auto &d) { d.build(input_filename, build_config); }, dict_);
+    if (!common::get_verbose())
+        std::cout.clear();
+
+    num_nodes_ = dict_size();
+}
+
+std::string DBGSSHash::file_extension() const {
+    return kExtension;
+}
+
+void DBGSSHash::add_sequence(std::string_view sequence,
+                             const std::function<void(node_index)>& on_insertion) {
+    throw std::logic_error("adding sequences not supported");
+}
+
+void DBGSSHash::map_to_nodes(std::string_view sequence,
+                             const std::function<void(node_index)>& callback,
+                             const std::function<bool()>& terminate) const {
+    if (mode_ != CANONICAL) {
+        map_to_nodes_sequentially(sequence, callback, terminate);
+    } else {
+        map_to_nodes_with_rc(
```

Comment:
```suggestion
        map_to_nodes_with_rc<true>(
```
To make it more explicit and for consistency with all other uses.

Reply:
already fixed :)